cpp_guidelines/guidelines.md

```diff
@@ -0,0 +1,496 @@
+# CRP & CASUS C++ Guidelines
+
+## General considerations
+
+### Lines are restricted to 79 columns.
+
+**Reason:** While 80 columns are a good value to view code side by side this
+does not take `diff` into account which adds another character.
+
+**Enforced by:**
+  * `.clang-format`: `ColumnLimit: 79`
+  * `.clang-format`: `BreakStringLiterals: true`
+  * `.clang-format`: `PenaltyBreakAssignment: 2`
+  * `.clang-format`: `PenaltyBreakBeforeFirstCallParameter: 19`
+  * `.clang-format`: `PenaltyBreakComment: 300`
+  * `.clang-format`: `PenaltyBreakFirstLessLess: 120`
+  * `.clang-format`: `PenaltyBreakString: 1000`
+  * `.clang-format`: `PenaltyBreakTemplateDeclaration: 10`
+  * `.clang-format`: `PenaltyExcessCharacter: 1000000`
+  * `.clang-format`: `PenaltyReturnTypeOnItsOwnLine: 1000`
+  * `.clang-format`: `ReflowComments: true`
+
+### Use UNIX-style linebreaks.
+
+**Reason:** Some editors interpret CRLF (`\r\n`) as two linebreaks. LF (`\n`)
+is always interpreted correctly.
+
+**Enforced by:**
+  * `.clang-format`: `DeriveLineEnding: false`
+  * `.clang-format`: `UseCRLF: false`
```

Comment:
I actually never had a problem with line endings, because git handles this automatically. So I wonder if we need to set anything here. I am afraid that it could mess up when formatting on Windows machines.

Reply:
Line endings depends on you git configuration. IMO if we enforce it we will avoid issues.

Reply:
So far this does not seem to be an issue on Windows.

Reply:
+1

Reply:
Ok, this turned out to be an issue for Windows now. Given a file with sections where clang-format is disabled:
```
    ...
    // clang-format off
    ...
    // clang-format on
    ...
```
Git by default will checkout with CRLF on Windows as it should. When you clang-format now, all lines are switched to LF, EXCEPT the lines where clang-format is disabled. This results in a file with mixed line endings and some IDEs rightfully complain now. Formatting the entire file using LF solves the IDE problem but then shows the entire file as changed in git.

Reply:
I am now wondering if we shouldn't enforce the correct line ending through git rather than clang-format. This should prevent the issues described by Bernhard.